File: Data_Structure/HW5/MyLinkedList.java
package src;

public class MyLinkedList<T> {
    LinkedListNode<T> head; //dummy head
    int numItems;

    MyLinkedList() {
        head = new LinkedListNode<>(null);
    }

    public int size() {
        return numItems;
    }

    public MyLinkedListIterator<T> getIterator(){return new MyLinkedListIterator<>(this);}

    public LinkedListNode<T> first() {
        return head.getNext();
    }

    public void insert(T item) {
        LinkedListNode<T> last = head;
        while (last.getNext() != null) {
            last = last.getNext();
        }
        last.insertNext(item);
        numItems += 1;
    }

    public MyLinkedList<T> copy() {
        MyLinkedList<T> cloneList = new MyLinkedList<>();
        LinkedListNode<T> cloneLast = cloneList.head;
        LinkedListNode<T> last = head;
        while (last.getNext() != null) {
            last = last.getNext();
            cloneLast.insertNext(last.getItem());
            cloneLast = cloneLast.getNext();
            cloneList.numItems++;
        }
        return cloneList;
    }

    static class LinkedListNode<T> {
        private LinkedListNode<T> next;
        private final T item;

        public LinkedListNode(T obj) {
            this.item = obj;
            this.next = null;
        }

        public T getItem() {
            return item;
        }

        public void setNext(LinkedListNode<T> next) {
            this.next = next;
        }

        public LinkedListNode<T> getNext() {
            return this.next;
        }

        public void removeNext() {
            setNext(next.getNext());
        }

        public void insertNext(T item) {
            setNext(new LinkedListNode<>(item));
        }
    }


    static class MyLinkedListIterator<T> {
        private final MyLinkedList<T> list;
        private LinkedListNode<T> curr;
        private LinkedListNode<T> prev;

        public LinkedListNode<T> getCurr() {
            return curr;
        }

        public MyLinkedListIterator(MyLinkedList<T> list) {
            this.list = list;
            this.curr = list.first();
            this.prev = list.head;
        }

        public void next() {
            prev = curr;
            curr = curr.getNext();
        }

        public void remove() {
            prev.removeNext();
            list.numItems--;
            curr = prev;
            prev = null;
        }
    }
}

File: Data_Structure/HW5/src/MyLinkedList.kt
package src

import src.AVL.AVLNode
import src.MyLinkedList
import src.MyHashtable
import kotlin.Throws
import java.io.IOException
import java.io.FileReader
import java.io.BufferedReader
import src.AVL
import src.MyLinkedList.MyLinkedListIterator
import src.MyLinkedList.LinkedListNode
import src.Database
import kotlin.jvm.JvmStatic
import src.Matching
import java.lang.NumberFormatException

class MyLinkedList<T> internal constructor() {
    var head //dummy head
            : LinkedListNode<T?>
    var numItems = 0
    fun size(): Int {
        return numItems
    }

    val iterator: MyLinkedListIterator<T>
        get() = MyLinkedListIterator(this)

    fun first(): LinkedListNode<T?>? {
        return head.next
    }

    fun insert(item: T) {
        var last: LinkedListNode<T?>? = head
        while (last!!.next != null) {
            last = last.next
        }
        last.insertNext(item)
        numItems += 1
    }

    fun copy(): MyLinkedList<T?> {
        val cloneList = MyLinkedList<T?>()
        var cloneLast: LinkedListNode<T?>? = cloneList.head
        var last: LinkedListNode<T?>? = head
        while (last!!.next != null) {
            last = last.next
            cloneLast!!.insertNext(last!!.item)
            cloneLast = cloneLast.next
            cloneList.numItems++
        }
        return cloneList
    }

    class LinkedListNode<T>(val item: T) {
        var next: LinkedListNode<T>? = null
        fun removeNext() {
            next = next!!.next
        }

        fun insertNext(item: T) {
            next = LinkedListNode(item)
        }
    }

    class MyLinkedListIterator<T>(private val list: MyLinkedList<T>) {
        var curr: LinkedListNode<T>?
            private set
        private var prev: LinkedListNode<T>?
        operator fun next() {
            prev = curr
            curr = curr!!.next
        }

        fun remove() {
            prev!!.removeNext()
            list.numItems--
            curr = prev
            prev = null
        }

        init {
            curr = list.first()
            prev = list.head
        }
    }

    init {
        head = LinkedListNode(null)
    }
}